File: backend/app/schemas/favorite.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class FavoriteCreateRequest(BaseModel):
    article_id: str = Field(alias="articleId")
    item_type: str = Field(alias="itemType")
    item_id: str = Field(alias="itemId")
    snapshot: dict[str, Any]


class FavoriteResponse(BaseModel):
    id: str
    article_id: str = Field(serialization_alias="articleId")
    article_title: str = Field(serialization_alias="articleTitle")
    item_type: str = Field(serialization_alias="itemType")
    item_id: str = Field(serialization_alias="itemId")
    snapshot: dict[str, Any]
    created_at: datetime = Field(serialization_alias="createdAt")


class FavoriteListResponse(BaseModel):
    items: list[FavoriteResponse]

File: frontend/src/types/favorite.ts
import type { Phrase, Sentence, WordExplanation } from './article'

export type FavoriteItemType = 'word' | 'phrase' | 'sentence'

export type FavoriteSnapshot = Phrase | WordExplanation | Sentence

export type Favorite = {
  id: string
  articleId: string
  articleTitle: string
  itemType: FavoriteItemType
  itemId: string
  snapshot: FavoriteSnapshot
  createdAt: string
}

export type FavoriteListResponse = {
  items: Favorite[]
}

export type FavoriteCreateInput = {
  articleId: string
  itemType: FavoriteItemType
  itemId: string
  snapshot: FavoriteSnapshot
}
